spec/outputs/rabbitmq_spec.rb

```diff
@@ -69,4 +69,64 @@
     end
 
   end
+
+  # TODO: Redo all this
+  # This is a crazy test to fix an urgent bug. It tests that we actually to change the
+  # store exchange value when an exception happens
+  # This should be a small number of lines, but until we refactor we must jump through hoops
+  # to put the plugin in the proper state and extract the proper state to test
+  describe "retrying a publish" do
+    let(:settings) {
+      {
+        "host" => "localhost",
+        "exchange_type" => "topic",
+        "exchange" => "foo",
+        "key" => "%{foo}"
+      }
+    }
+    let(:event) { LogStash::Event.new("foo" => "bar")}
+
+    subject { LogStash::Outputs::RabbitMQ.new(settings) }
+
+    before do
+      allow(subject).to receive(:connect) {
+                          subject.instance_variable_get(:@connected).set(true)
+                        }
+
+      exchange_invocation = 0
+
+      @most_recent_exchange = nil
+
+      allow(subject).to receive(:declare_exchange) do
+        exchange_invocation += 1
+        @most_recent_exchange = double("exchange #{exchange_invocation}")
+      end
+
+      subject.register
+    end
+
+    context "when the connection aborts once" do
+      before do
+        i = 0
+        allow(subject).to receive(:attempt_publish_serialized) do
+          i+=1
+          if i == 1 # Raise an exception once
+           raise MarchHare::Exception, "First"
+          else
+            "nope"
+          end
+        end
+
+        subject.publish_serialized(event, "hello")
+      end
+
+      it "should re-declare the exchange" do
+        expect(subject).to have_received(:declare_exchange).twice
```

Comment:
The goal of this test is to ascertain that we create a new exchange twice. Once on `#register` once after we hit the exception. Hence, we check that it is invoked twice exactly, and not say, a superfluous third time.